# Advice

Here is general advice and tutorials

## Fungal phylogenomics

Bryn Dentinger has written a pipeline / wiki for fungal phylogenomics is available at [Phylogenomic tools wiki](Phylogenomic_tools_wiki.md).

## Commonly used software

* QUAST
 * Genome assembly stats calculator including shiny output in .pdf or .html
 * [Manual](http://quast.bioinf.spbau.ru/manual.html)
* SPAdes (and hybridSPAdes)
 * De novo short-read (SPAdes) and hybrid long/short-read (hybridSPAdes) genome assembly. 
 * [Manual](http://spades.bioinf.spbau.ru/release3.7.1/manual.html)
 
## Links to other tutorials

* [Analysing nanopore data with SPAdes](http://angus.readthedocs.io/en/stable/analyzing_nanopore_data.html)
* [De novo assembly with SPAdes](http://angus.readthedocs.io/en/stable/assembling-ecoli.html)
* [Trimming and mapping reads to a genome with FastQC, trimGalore, BWA](https://www.biostars.org/p/75489/)

