# **Analyse target capture data to obtain gene alignments**
  
## **1. Working environment and best practices**

Processing sequencing data is much easier, faster and customizable using command line. Many bioinformatic tools are free and designed to work in a unix-like environment, so people usually analyze NGS data on OS X or Linux computers.  
  
There are already many free tools, pipelines and tutorials available so most often you just need to know enough basic commands to link these tools together and to parse results, so don’t panic, and don’t hesitate to google your problem!  
  
More important than knowing many commands is to understand the principles behind the tool that you use, so that you can correctly analyze the data and properly interpret the results. But this does not require high bioinformatic skills, it requires to read the manual or article associated with the tool. More on this topic is provided here.  
  
When you cannot find the tool you need or when you have difficulties to understand what is happening, ask a (bio)informatician!  
  
Kew has computing and High Performance Computing (HPC) facilities that have been set up to enable the analysis of large amounts of target capture sequencing data by multiple users. Before using them, please ensure that:  
  
- you make yourself known to Kew’s (bio)informaticians, 
- you get familiar with the facilities, 
- you take note of the best practices described here.

## **2. Checking data quality**
## **3. Cleaning data**
